contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,582 @@
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+import { IBasicIssuanceModule } from "../interfaces/IBasicIssuanceModule.sol";
+import { IDebtIssuanceModule } from "../interfaces/IDebtIssuanceModule.sol";
+import { IController } from "../interfaces/IController.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { IWETH } from "../interfaces/IWETH.sol";
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+
+
+contract ExchangeIssuanceZeroEx is Ownable, ReentrancyGuard {
+
+    using Address for address payable;
+    using SafeMath for uint256;
+    using PreciseUnitMath for uint256;
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ISetToken;
+
+    struct ZeroExSwapQuote {
+        IERC20 sellToken;
+        IERC20 buyToken;
+        bytes swapCallData;
+    }
+
+    struct IssuanceModuleData {
+        bool isAllowed;
+        bool isDebtIssuanceModule;
+    }
+
+    /* ============ Constants ============== */
+
+    address constant public ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /* ============ State Variables ============ */
+
+    address public immutable WETH;
+    mapping(address => IssuanceModuleData) public allowedIssuanceModules;
+
+    IController public immutable setController;
+
+    address public swapTarget;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,     // The recipient address of the issued SetTokens
+        ISetToken indexed _setToken,    // The issued SetToken
+        IERC20 indexed _inputToken,     // The address of the input asset(ERC20/ETH) used to issue the SetTokens
+        uint256 _amountInputToken,      // The amount of input tokens used for issuance
+        uint256 _amountSetIssued        // The amount of SetTokens received by the recipient
+    );
+
+    event ExchangeRedeem(
+        address indexed _recipient,     // The recipient address which redeemed the SetTokens
+        ISetToken indexed _setToken,    // The redeemed SetToken
+        IERC20 indexed _outputToken,    // The address of output asset(ERC20/ETH) received by the recipient
+        uint256 _amountSetRedeemed,     // The amount of SetTokens redeemed for output tokens
+        uint256 _amountOutputToken      // The amount of output tokens received by the recipient
+    );
+
+    event Refund(
+        address indexed _recipient,     // The recipient address which redeemed the SetTokens
+        uint256 _refundAmount           // The amount of ETH redunded to the recipient
+    );
+
+    event BoughtTokens(IERC20 sellToken, IERC20 buyToken, uint256 boughtAmount);
+
+    /* ============ Modifiers ============ */
+
+    modifier isSetToken(ISetToken _setToken) {
```

Comment:
I think we don't need this either. Every issuance and redemption will check if  it is a valid set token anyway so it is redundant The only time where it can get weird is in `approveSetToken`, but I don't see any malicious activity that could be done by passing in a fake set token.

Reply:
Agreed. We do call state changing methods on this address (`approve`) but I guess there is no reentrancy risk or anything from this.

Removed the modifier accordingly